File: app/src/main/java/com/example/vgcamera/FaceAnalyzer.java
package com.example.vgcamera;

import android.content.Context;
import android.graphics.ImageFormat;
import android.media.Image;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.camera.core.ImageAnalysis;
import androidx.camera.core.ImageProxy;

import com.google.mlkit.vision.common.InputImage;
import com.google.mlkit.vision.face.Face;
import com.google.mlkit.vision.face.FaceDetection;
import com.google.mlkit.vision.face.FaceDetector;
import com.google.mlkit.vision.face.FaceDetectorOptions;

import java.nio.ByteBuffer;
import java.util.List;

public class FaceAnalyzer implements ImageAnalysis.Analyzer {

    public interface FaceDetectionListener {
        void onFaceLookingStraight();
        void onFaceNotLookingStraight();
        void onFaceAreaLarge(boolean isLarge);
        void onAverageLuminance(double luminance);  // Thêm callback nhận độ sáng trung bình
    }

    private final FaceGraphicOverlay overlay;
    private final FaceDetectionListener listener;
    private final FaceDetector detector;
    private final Context context;

    public FaceAnalyzer(Context context, FaceGraphicOverlay overlay, FaceDetectionListener listener) {
        this.context = context;
        this.overlay = overlay;
        this.listener = listener;

        FaceDetectorOptions options = new FaceDetectorOptions.Builder()
                .setPerformanceMode(FaceDetectorOptions.PERFORMANCE_MODE_FAST)
                .build();

        detector = FaceDetection.getClient(options);
    }

    @Override
    public void analyze(@NonNull ImageProxy imageProxy) {
        int detectThreshold = context.getSharedPreferences("settings", Context.MODE_PRIVATE)
                .getInt("detect_threshold", 3000);

        int recognizeThreshold = context.getSharedPreferences("settings", Context.MODE_PRIVATE)
                .getInt("recognize_threshold", 3000);

        Log.e("detect_threshold", String.valueOf(detectThreshold));
        Log.e("recognize_threshold", String.valueOf(recognizeThreshold));

        Image mediaImage = imageProxy.getImage();

        if (mediaImage == null) {
            imageProxy.close();
            return;
        }

        // --- Tính độ sáng trung bình từ kênh Y (YUV_420_888) ---
        if (mediaImage.getFormat() == ImageFormat.YUV_420_888) {
            ByteBuffer buffer = mediaImage.getPlanes()[0].getBuffer(); // Lấy plane Y
            byte[] data = new byte[buffer.remaining()];
            buffer.get(data);

            long sum = 0;
            for (byte b : data) {
                sum += (b & 0xFF);
            }
            double avgLuminance = sum / (double) data.length;

            if (listener != null) {
                listener.onAverageLuminance(avgLuminance);
            }
        }

        InputImage image = InputImage.fromMediaImage(mediaImage, imageProxy.getImageInfo().getRotationDegrees());

        detector.process(image)
                .addOnSuccessListener(faces -> {
                    if (overlay != null) {
                        List<Face> filteredFaces = new java.util.ArrayList<>();
                        for (Face face : faces) {
                            float faceArea = face.getBoundingBox().width() * face.getBoundingBox().height();
                            if (faceArea > detectThreshold) {
                                filteredFaces.add(face);
                            }
                        }
                        overlay.setFaces(filteredFaces);
                    }

                    boolean foundStraightAndCloseFace = false;
                    boolean foundLargeFace = false;

                    for (Face face : faces) {
                        float rotY = face.getHeadEulerAngleY();
                        float rotZ = face.getHeadEulerAngleZ();

                        float faceArea = face.getBoundingBox().width() * face.getBoundingBox().height();

                        if (faceArea > detectThreshold) {
                            foundLargeFace = true;
                        }

                        if (Math.abs(rotY) < 10 && Math.abs(rotZ) < 10 && faceArea > recognizeThreshold) {
                            foundStraightAndCloseFace = true;
                        }
                    }

                    listener.onFaceAreaLarge(foundLargeFace);

                    if (foundStraightAndCloseFace) {
                        listener.onFaceLookingStraight();
                    } else {
                        listener.onFaceNotLookingStraight();
                    }
                })
                .addOnFailureListener(e -> Log.e("FaceAnalyzer", "Face detection failed", e))
                .addOnCompleteListener(task -> imageProxy.close());
    }
}

File: app/src/main/java/com/example/vgcamera/FaceAnalyzer.kt
package com.example.vgcamera

import android.content.Context
import android.graphics.ImageFormat
import android.util.Log
import androidx.camera.core.ImageAnalysis
import androidx.camera.core.ImageProxy
import com.example.vgcamera.FaceAnalyzer.FaceDetectionListener
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.face.Face
import com.google.mlkit.vision.face.FaceDetection
import com.google.mlkit.vision.face.FaceDetectorOptions
import java.util.ArrayList
import kotlin.math.abs

import com.google.mlkit.vision.facemesh.FaceMesh
import com.google.mlkit.vision.facemesh.FaceMeshDetection
import com.google.mlkit.vision.facemesh.FaceMeshDetector
import com.google.mlkit.vision.facemesh.FaceMeshDetectorOptions

class FaceAnalyzer(
    private val context: Context,
    private val overlay: FaceGraphicOverlay?,
    private val listener: FaceDetectionListener?
) : ImageAnalysis.Analyzer {

    interface FaceDetectionListener {
        fun onFaceMeshDetected(mesh: FaceMesh, imageWidth: Int, imageHeight: Int, rotation: Int, isLookingStraight: Boolean)
        fun onNoFaceDetected()
        fun onFaceInTarget(mesh: FaceMesh, isLookingStraight: Boolean)
        fun onFaceNotInTarget()
        fun onAverageLuminance(luminance: Double)
    }

    private val detector: FaceMeshDetector by lazy {
        FaceMeshDetection.getClient(
            FaceMeshDetectorOptions.Builder()
                .build()
        )
    }

    override fun analyze(imageProxy: ImageProxy) {
        val prefs = context.getSharedPreferences("settings", Context.MODE_PRIVATE)
        val recognizeThreshold = prefs.getInt("recognize_threshold", 5000)

        val mediaImage = imageProxy.image ?: run {
            imageProxy.close()
            return
        }

        val rotation = imageProxy.imageInfo.rotationDegrees

        // --- Average Luminance ---
        if (mediaImage.format == ImageFormat.YUV_420_888) {
            val buffer = mediaImage.planes[0].buffer
            val data = ByteArray(buffer.remaining())
            buffer.get(data)
            var sum: Long = 0
            for (b in data) sum += (b.toInt() and 0xFF).toLong()
            listener?.onAverageLuminance(sum.toDouble() / data.size)
        }

        val image = InputImage.fromMediaImage(mediaImage, rotation)

        detector.process(image)
            .addOnSuccessListener { meshes ->
                processMeshes(meshes, recognizeThreshold, image.width, image.height, rotation)
            }
            .addOnFailureListener { e ->
                Log.e("FaceAnalyzer", "Face Mesh detection failed", e)
                listener?.onNoFaceDetected()
            }
            .addOnCompleteListener {
                imageProxy.close()
            }
    }

    private fun processMeshes(
        meshes: List<FaceMesh>,
        recognizeThreshold: Int,
        width: Int,
        height: Int,
        rotation: Int
    ) {
        if (meshes.isEmpty()) {
            listener?.onNoFaceDetected()
            listener?.onFaceNotInTarget()
            return
        }

        var mainMesh: FaceMesh? = null
        var maxArea = -1f
        for (mesh in meshes) {
            val area = (mesh.boundingBox.width() * mesh.boundingBox.height()).toFloat()
            if (area > maxArea) {
                maxArea = area
                mainMesh = mesh
            }
        }

        if (mainMesh == null) {
            listener?.onNoFaceDetected()
            listener?.onFaceNotInTarget()
            return
        }

        // --- 🔎 KIỂM TRA NHÌN THẲNG (SYMMETRY CHECK) ---
        // Sử dụng các điểm mốc: Mũi (6), Mắt trái (133), Mắt phải (362)
        val pts = mainMesh.allPoints
        var isStraight = false
        try {
            val nose = pts[6].position
            val leftEye = pts[133].position
            val rightEye = pts[362].position
            
            val distL = abs(nose.x - leftEye.x)
            val distR = abs(rightEye.x - nose.x)
            val ratio = if (distL > distR) distR / distL else distL / distR
            
            // Nếu ratio > 0.7 và độ lệch Y của mắt nhỏ -> Nhìn khá thẳng
            val eyeDiffY = abs(leftEye.y - rightEye.y)
            isStraight = ratio > 0.75f && eyeDiffY < (mainMesh.boundingBox.height() * 0.05f)
        } catch (e: Exception) {
            isStraight = true // Fallback
        }

        val isRotated = rotation == 90 || rotation == 270
        val analyzedW = if (isRotated) height else width
        val analyzedH = if (isRotated) width else height

        val box = mainMesh.boundingBox
        val centerX = box.centerX()
        val centerY = box.centerY()

        val safetyMarginW = analyzedW * 0.20
        val safetyMarginH = analyzedH * 0.20

        val isInTarget = centerX > safetyMarginW && centerX < (analyzedW - safetyMarginW) &&
                         centerY > safetyMarginH && centerY < (analyzedH - safetyMarginH) &&
                         maxArea > recognizeThreshold

        listener?.onFaceMeshDetected(mainMesh, analyzedW, analyzedH, rotation, isStraight)

        if (isInTarget) {
            listener?.onFaceInTarget(mainMesh, isStraight)
        } else {
            listener?.onFaceNotInTarget()
        }
    }
}
